### ISSUE TITLE
dependencies cannot be imported if the package name differs from the lib name in the depenency and type is cdylib

### ISSUE THREAD

**Issue by paul-callahan** [rust-lang/cargo#4231]
The cargo build will fail a crate is attempted to be imported with `extern crate ...` for a dependency whose [package] name differs from the name in the [lib] section and crate-type is set to ["cdylib"].

Here is a reproduction.   Note `libfoo` is the dependency and `bar` is the thing trying to use `libfoo`.   In the "package" section of `libfoo`, name is "libfoo" and in the [lib] section name is "foo"

libfoo/Cargo.toml
```
[package]
name = "libfoo"
version = "0.1.0"

[dependencies]

[lib]
name = "foo"
crate-type = ["cdylib"]
```

being used in this project:
bar/Cargo.toml
```
[package]
name = "bar"
version = "0.1.0"

[dependencies]
libfoo = { path = "/tmp/libfoo" }

[lib]
name = "bar"
crate-type = ["cdylib"]
```

This on its own compiles ok.   But now try to import it into `bar/src/lib.rs`:

```
extern crate libfoo;   //also foo doesn't work.

#[cfg(test)]
mod tests {
    #[test]
    fn it_works() {
    }
}
```

Cargo fails with this error:
```
error[E0463]: can't find crate for `libfoo`
 --> src/lib.rs:1:1
  |
1 | extern crate libfoo;
  | ^^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error(s)

error: Could not compile `bar`.
```

Now if you delete the entire `[lib]` section in `libfoo/Cargo.toml` ,compilation works.   Or if you remove the `crate-type` and use `extern crate foo`, it compilation works as well.

**Comment by paul-tcell** [comment]
[repro.zip](https://github.com/rust-lang/cargo/files/1103793/repro.zip)

**Comment by alexcrichton** [comment]
@paul-callahan ah this is expected behavior because a `cdylib` can't be loaded via `extern crate`, only a `dylib` or an `rlib` can be loaded as such. In any case Cargo's error message shoudl be better here.

**Comment by dwijnand** [comment]
Is this a sanity check that cargo should do, or one rustc should be responsible for?

**Comment by alexcrichton** [comment]
Unfortunately it's somewhat unclear where this would go. We allow depending on crates with cdylib/bin/etc I think largely for backwards-compatibility with old mistakes, but Cargo doesn't actually know when you depend on a package whether you'll actually attempt to load its crate or not. The fix here is probably going to be twofold between Cargo/rustc where Cargo tells rustc "if this crate is requestd it's denied because it doesn't have an rlib", and then if rustc actually attempts to load such a crate it'll defer to cargo for the error message (or something like that)

Not an easy bug to fix :(

**Comment by eminence** [comment]
Triage:  cargo now gives a more helpful error message when building:

> warning: The package `foo` provides no linkable target. The compiler might raise an error while compiling `bar`. Consider adding 'dylib' or 'rlib' to key `crate-type` in `foo`'s Cargo.toml. This warning might turn into a hard error in the future.

This message was implemented in #4809 as a fix for #3169

**Comment by epage** [comment]
Now that we have a warning, it looks like this is as resolved as it can be so I'm going to close this.  If there is a reason we should re-evaluate this, let us know!
